pkg/gcrane/copy.go

```diff
@@ -144,83 +147,131 @@ func copyIndex(src, dst string, srcAuth, dstAuth authn.Authenticator) error {
 		return fmt.Errorf("parsing reference %q: %v", dst, err)
 	}
 
-	idx, err := remote.Index(srcRef, remote.WithAuth(srcAuth))
+	idx, err := remote.Index(srcRef, remote.WithAuth(srcAuth), remote.WithTransport(t))
 	if err != nil {
 		return fmt.Errorf("reading image %q: %v", src, err)
 	}
 
-	if err := remote.WriteIndex(dstRef, idx, dstAuth, http.DefaultTransport); err != nil {
+	if err := remote.WriteIndex(dstRef, idx, dstAuth, t); err != nil {
 		return fmt.Errorf("writing image %q: %v", dst, err)
 	}
 
 	return nil
 }
 
-// recursiveCopy copies images from repo src to repo dst, rather quickly. tl;dr:
-//
-//  for each repo in src {
-//		go func {
-//			for each image in repo {
-//				go func {
-//					for each tag in image {
-//						go func {
-//							copyImage(tag, rename(tag, dst))
-//						}
-//					}
-//				}
-//			}
-//			for each index in repo {
-//				go func {
-//					for each tag in index {
-//						go func {
-//							copyIndex(tag, rename(tag, dst))
-//						}
-//					}
-//				}
-//			}
-//		}
-//	}
+type throttledTransport struct {
```

Comment:
It seems a bit weird to throttle at the transport level, is that just because it's the easiest/lowest place to inject the wait?

It might make more sense to have the `rate.Wait` [in the loop where goroutines are pulling tasks from the `tasks` chan](https://github.com/google/go-containerregistry/blob/4c6c546743b750542d3aa6fb224ad533e6aa404a/pkg/gcrane/copy.go#L240), but then it's not enforcing HTTP-requests-per-second, it's limiting copy-operations-per-second, where each might involve multiple HTTP requests.

A more robust solution would be to have the transport backoff and retry when it gets a 429, and then we don't have to guess what the rate limit will be ahead of time, we can just do as much as we can until the server tells us to back off.

Reply:
> It seems a bit weird to throttle at the transport level, is that just because it's the easiest/lowest place to inject the wait?

I think it's the most correct. GCR will throttle at the request level, so this is pretty close to the truth. I think we may be overcounting some requests (e.g. clients get redirected to GCS, we'll count the GCS request and the GCR redirect twice).

> A more robust solution would be to have the transport backoff and retry when it gets a 429, and then we don't have to guess what the rate limit will be ahead of time, we can just do as much as we can until the server tells us to back off.

 I think this is fine but we will effectively do a ton of work at once, then wait ten minutes before continuing. For interactive use cases that's not much better than failing (IMO), but for non-interactive cases this is probably better than doing any throttling... good point. I will update to do this.